### PR TITLE
refactor: pull kepubify & ffprobe during build

### PR DESCRIPTION
**Pull request by imnotjames**
## 📝 Description

The kepubify and ffprobe binaries are currently pulled at runtime from a github repo.  If we instead include them during the docker build, we no longer need to maintain a separate repository of just the binaries.

This also means that updating the kepubify / ffprobe can be done per-Grimmory version rather than "for everyone anywhere irregardless of what version they are using" which improves compatibility.

## 🏷️ Type of Change

- [ ] Bug fix
- [ ] New feature
- [ ] Enhancement to existing feature
- [x] Refactor (no behavior change)
- [ ] Breaking change (existing functionality affected)
- [ ] Documentation update

## 🔧 Changes

- Create a helper in fileservice to search system path + local tools folder
- Update the kepubify service to use a local kepubify if available
- Drop kepubify fetch during runtime
- Update the ffprobe service to use a local ffprobe if available
- Drop ffprobe fetch during runtime 
- Update the dockerfile to include kepubify & ffprobe during build

## 🧪 Testing (MANDATORY)

> **PRs without this section filled out will be closed.** "Tests pass" or "Tested locally" is not sufficient. You must provide specifics.

**Manual testing steps you performed:**

1. Start dev compose
2. Turn on kepubify option & kobo sync
3. Make request against `localhost:6060/api/kobo/{key}/v1/books/1/download` and get file
4. Observe file in kepub format

Testing the audiobook feature was more difficult as it's a fallback to use ffprobe.  However, I did try deleting the "normal" ffprobe path and it seemed to still retrieve chapters?

**Regression testing:**

- Ran kepubify, no issues

**Edge cases covered:**

- Missing file in path
- kepubify in local bin
- kepubify in multiple locations (selects first in order)

**Test output:**

<details>
<summary>Backend test output (<code>./gradlew test</code>)</summary>

```
> Task :compileTestJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
WARNING: A terminally deprecated method in sun.misc.Unsafe has been called
WARNING: sun.misc.Unsafe::objectFieldOffset has been called by net.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe$Dispatcher$CreationAction (file:/home/james/.local/share/gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy/1.17.8/af5735f63d00ca47a9375fae5c7471a36331c6ed/byte-buddy-1.17.8.jar)
WARNING: Please consider reporting this to the maintainers of class net.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe$Dispatcher$CreationAction
WARNING: sun.misc.Unsafe::objectFieldOffset will be removed in a future release
2026-03-19T13:18:26.708-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-19T13:18:26.708-04:00  INFO 1342130 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-19T13:18:26.802-04:00  INFO 1342130 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-19T13:18:26.803-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.LibraryWatchService     : Shutting down LibraryWatchService...
2026-03-19T13:18:26.804-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-19T13:18:26.804-04:00  WARN 1342130 --- [booklore-api] [     watch-poll] o.b.s.monitoring.LibraryWatchService     : WatchService closed, stopping poll loop.
2026-03-19T13:18:26.815-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-19T13:18:26.834-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-19T13:18:26.835-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-19T13:18:26.843-04:00  INFO 1342130 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-19T13:18:26.843-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-19T13:18:26.843-04:00  INFO 1342130 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-19T13:18:26.844-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.LibraryWatchService     : Shutting down LibraryWatchService...
2026-03-19T13:18:26.844-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-19T13:18:26.844-04:00  WARN 1342130 --- [booklore-api] [     watch-poll] o.b.s.monitoring.LibraryWatchService     : WatchService closed, stopping poll loop.
2026-03-19T13:18:26.850-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-19T13:18:26.851-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-19T13:18:26.852-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-19T13:18:26.858-04:00  INFO 1342130 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-19T13:18:26.859-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-19T13:18:26.859-04:00  INFO 1342130 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-19T13:18:26.861-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.LibraryWatchService     : Shutting down LibraryWatchService...
2026-03-19T13:18:26.861-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-19T13:18:26.861-04:00  WARN 1342130 --- [booklore-api] [     watch-poll] o.b.s.monitoring.LibraryWatchService     : WatchService closed, stopping poll loop.
2026-03-19T13:18:26.872-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-19T13:18:26.886-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-19T13:18:26.886-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-19T13:18:26.894-04:00  INFO 1342130 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-19T13:18:26.898-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-19T13:18:26.898-04:00  INFO 1342130 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-19T13:18:26.900-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.LibraryWatchService     : Shutting down LibraryWatchService...
2026-03-19T13:18:26.900-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-19T13:18:26.900-04:00  WARN 1342130 --- [booklore-api] [     watch-poll] o.b.s.monitoring.LibraryWatchService     : WatchService closed, stopping poll loop.
2026-03-19T13:18:26.911-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-19T13:18:26.921-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-19T13:18:26.921-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-19T13:18:26.929-04:00  INFO 1342130 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-19T13:18:26.929-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-19T13:18:26.929-04:00  INFO 1342130 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-19T13:18:26.931-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.LibraryWatchService     : Shutting down LibraryWatchService...
2026-03-19T13:18:26.931-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-19T13:18:26.931-04:00  WARN 1342130 --- [booklore-api] [     watch-poll] o.b.s.monitoring.LibraryWatchService     : WatchService closed, stopping poll loop.
2026-03-19T13:18:26.942-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-19T13:18:26.950-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-19T13:18:26.950-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-19T13:18:26.953-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopping...
2026-03-19T13:18:26.953-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : BrokerAvailabilityEvent[available=false, SimpleBrokerMessageHandler [org.springframework.messaging.simp.broker.DefaultSubscriptionRegistry@4937128b]]
2026-03-19T13:18:26.953-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopped.
2026-03-19T13:18:26.958-04:00  INFO 1342130 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-19T13:18:26.959-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-19T13:18:26.959-04:00  INFO 1342130 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-19T13:18:26.961-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.LibraryWatchService     : Shutting down LibraryWatchService...
2026-03-19T13:18:26.961-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-19T13:18:26.962-04:00  WARN 1342130 --- [booklore-api] [     watch-poll] o.b.s.monitoring.LibraryWatchService     : WatchService closed, stopping poll loop.
2026-03-19T13:18:26.971-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-19T13:18:26.978-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-19T13:18:26.979-04:00  INFO 1342130 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.

[Incubating] Problems report is available at: file:///home/james/Projects/imnotjames/grimmory/booklore-api/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.4.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2m 11s
6 actionable tasks: 3 executed, 3 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.4.0/userguide/configuration_cache_enabling.html
┌─[13:18:29]─[james at ffffffff]─[fix/-/kepubify-and-ffprobe]

```

</details>

<details>
<summary>Frontend test output (<code>ng test</code>)</summary>

N/A

</details>

## 📸 Screen Recording / Screenshots (MANDATORY)

> Every PR must include a **screen recording or screenshots** showing the change working end-to-end in a running local instance (both backend and frontend). This means you must have actually built, run, and tested the code yourself. PRs without visual proof will be closed without review.

Not sure what I should take a picture of to prove the API is operating as expected.

---

## ✅ Pre-Submission Checklist

> **All boxes must be checked before requesting review.** Incomplete PRs will be closed without review. No exceptions.

- [ ] This PR is linked to an approved issue
- [x] Code follows project [backend and frontend conventions](../CONTRIBUTING.md#backend-conventions)
- [x] Branch is up to date with `develop` (merge conflicts resolved)
- [x] I ran the full stack locally (backend + frontend + database) and verified the change works
- [x] Automated tests added or updated to cover changes (backend **and** frontend)
- [x] All tests pass locally and output is pasted above
- [ ] Screen recording or screenshots are attached above proving the change works
- [x] PR is a single focused change (one bug fix OR one feature, not multiple unrelated changes)
- [x] PR is reasonably scoped (PRs over 1000+ changed lines will be closed, split into smaller PRs)
- [x] No unsolicited refactors, cleanups, or "improvements" are bundled in
- [x] Flyway migration versioning is correct _(if schema was modified)_
- [ ] Documentation PR submitted to [booklore-docs](https://github.com/booklore-app/booklore-docs) _(if user-facing changes)_

### 🤖 AI-Assisted Contributions

> **If any part of this PR was generated or assisted by AI tools (Copilot, Claude, ChatGPT, etc.), all items below are mandatory.** You are fully responsible for every line you submit. "The AI wrote it" is not an excuse, and AI-generated PRs that clearly haven't been reviewed are the #1 reason PRs get closed.

No AI was used.
---

## Summary by CodeRabbit

* **Refactor**
  * Centralized system binary discovery across the application, improving how external tools are located and resolved.

* **Chores**
  * Updated Docker image to include pre-built architecture-specific binaries for `kepubify` and `ffprobe`, eliminating runtime downloads.

### CODE REVIEW

File: booklore-api/src/main/java/org/booklore/service/kobo/KepubConversionService.java [hunk 1]
```diff
@@ -19,19 +19,15 @@ public class KepubConversionService {
...
-    private static final String BIN_LINUX_ARM64 = "kepubify-linux-arm64";
-
     public File convertEpubToKepub(File epubFile, File tempDir, boolean forceEnableHyphenation) throws IOException, InterruptedException {
         validateInputs(epubFile);
 
-        Path kepubifyBinary = setupKepubifyBinary();
+        Path kepubifyBinary = fileService.findSystemFile("kepubify");
+
+        if (kepubifyBinary == null) {
+            throw new IOException("Kepubify conversion failed: could not find kepubify binary");
```
Comment (imnotjames) [inline booklore-api/src/main/java/org/booklore/service/kobo/KepubConversionService.java:25]: Without this, the `null` is passed in and a different less clear failure occurs (NPE)

File: booklore-api/src/main/java/org/booklore/util/FileService.java [hunk 2]
```diff
@@ -148,12 +148,20 @@ public String getTempBookdropCoverImagePath(long bookdropFileId) {
...
+        // If not found in those, then search the system $PATH.
+        String[] searchPaths = "bin:tools:".concat(System.getenv("PATH")).split(":");
+
+        for (String path : searchPaths) {
+            Path possiblePath = Paths.get(path).resolve(filename).toAbsolutePath();
+
+            if (Files.exists(possiblePath)) {
+                return possiblePath;
+            }
+        }
```
Comment (imnotjames) [inline booklore-api/src/main/java/org/booklore/util/FileService.java:164]: If there's a better way to do this I'm open to hearing it. This seemed to be the simplest and most reliable.

File: Dockerfile [hunk 3]
```diff
@@ -37,6 +37,27 @@ RUN apk add --no-cache yq && \
...
+ARG KEPUBIFY_VERSION="4.0.4"
+ARG KEPUBIFY_AMD64_CHECKSUM="sha256:37d7628d26c5c906f607f24b36f781f306075e7073a6fe7820a751bb60431fc5"
+ARG KEPUBIFY_ARM64_CHECKSUM="sha256:5a15b8f6f6a96216c69330601bca29638cfee50f7bf48712795cff88ae2d03a3"
+
+ADD \
+      --checksum="${KEPUBIFY_AMD64_CHECKSUM}" \
+      https://github.com/pgaskin/kepubify/releases/download/v${KEPUBIFY_VERSION}/kepubify-linux-64bit /kepubify-amd64
+ADD \
+      --checksum="${KEPUBIFY_ARM64_CHECKSUM}" \
+      https://github.com/pgaskin/kepubify/releases/download/v${KEPUBIFY_VERSION}/kepubify-linux-arm64 /kepubify-arm64
```
Comment (imnotjames) [inline Dockerfile]: Given these binaries are small, this was the easiest way I could translate from the `kepubify` arch definitions and $TARGETARCH -- other options would require extra arguments or some more in depth shell scripting.

File: booklore-api/src/main/java/org/booklore/util/FileService.java (hunk 2 above)
Comment (balazs-szucs) [reply booklore-api/src/main/java/org/booklore/util/FileService.java:164]: I feel like it shouldn't be a big deal. Only danger here are Proxmox/TrueNAS installs, who are unofficial install scripts, so who knows how those scripts are setup. But I think both of those projects are reasonable enough not mess with the paths hardcore.
